Thanks for tracking this down. The bug reproduces for me: with two PIDs, `kill "$port_num"` gets a single argument containing a newline and fails. Narrowing with `-sTCP:LISTEN` looks right too.

Two things I ran into while reading it.

**The loop can report failure on a run that worked.** For a forked server, such as Puma or Node in cluster mode, the parent and every worker inherit the listening socket, so they all appear in the LISTEN list. The loop signals the parent first. The parent then tears down its workers, and by the time the loop reaches the worker PIDs they are already gone. `kill` writes `No such process` for each one and returns 1, and that becomes the exit status of the script. Passing every PID to a single `kill` avoids this, because all the signals go out before anything can be reaped.

**The pull request description does not match the code.** The description says "Leave the expansion unquoted so that each PID becomes an argument of its own", which is the batched version. The diff uses the loop instead. The commit body is accurate, so it is only the description that has drifted.

The unquoted form does trip shellcheck (SC2086). Piping into `xargs` gives you both the batched signals and a clean lint:

```sh
lsof -ti4TCP:"$1" -sTCP:LISTEN | xargs -r kill
```

I checked that with shellcheck 0.11.0 and it is clean. With a stubbed `lsof`: no match exits 0, two dead PIDs give one error line each and exit 1, and a missing argument still prints the usage and exits 1.

On `-r`, the BSD `xargs` on macOS accepts the flag, and it does not run the command on empty input anyway. GNU `xargs` runs the command once on empty input unless you pass `-r`. I did not have GNU findutils to hand, so that half is from the documentation rather than something I ran.

One last thing worth folding in: the header comment still says "Kills the process running on the provided port", singular, which is the assumption this change exists to correct.

--- a/bin/clear-port
+++ b/bin/clear-port
@@ -1,14 +1,13 @@
 #!/bin/sh
 
-# Kills the process running on the provided port
+# Kills the processes listening on the provided port
 #
 # clear-port 3000
 
 if [ -n "$1" ]; then
-  port_num="$(lsof -ti4TCP:"$1")"
-  if [ $? -eq 0 ]; then
-    kill "$port_num"
-  fi
+  # A port can be held by more than one process, and a forked server's workers
+  # die with the parent, so signal every PID in a single kill.
+  lsof -ti4TCP:"$1" -sTCP:LISTEN | xargs -r kill
 else
   echo >&2 Usage: clear-port port-number
   exit 1

